pkg/querier/worker/scheduler_processor.go

```diff
@@ -136,20 +137,45 @@ func (sp *schedulerProcessor) processQueriesOnSingleStream(workerCtx context.Con
 }
 
 // process loops processing requests on an established stream.
-func (sp *schedulerProcessor) querierLoop(c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) (err error) {
+func (sp *schedulerProcessor) querierLoop(execCtx context.Context, c schedulerpb.SchedulerForQuerier_QuerierLoopClient, address string, inflightQuery *atomic.Bool) (err error) {
 	// Build a child context so we can cancel a query when the stream is closed.
-	ctx, cancel := context.WithCancelCause(c.Context())
+	// Note that we deliberately don't use c.Context() here, as that is cancelled as soon as the gRPC client observes an error,
+	// but we don't always want to cancel queries if the scheduler stream reports an error (eg. if the scheduler crashed).
+	ctx, cancel := context.WithCancelCause(execCtx)
 	defer func() {
 		cancel(util.NewCancellationErrorf("query-scheduler loop in querier for query-scheduler %v terminated with error: %w", address, err))
 	}()
 
+	var queryComplete chan struct{}
+
+	waitForQuery := func(err error) {
+		select {
+		case <-queryComplete:
+			// Query is already complete, nothing to do.
+			return
+		default:
+			// Query is not complete.
+			level.Info(sp.log).Log("msg", "query-scheduler loop received non-cancellation error, waiting for inflight query to complete...", "err", err, "address", address)
```

Comment:
[nit] I think that calling it "query-scheduler loop" may be deceiving, because this code effectively runs in the querier. In the error above we say "query-scheduler loop in querier" I think is more clear. We could do the same here.

Reply:
Makes sense, changed in b5634d9